wise_pizza/solve/tree.py

```diff
@@ -0,0 +1,181 @@
+                    raise ValueError("NaNs in encoded values")
+                # Get split candidates for brute force search
+                deciles = np.array([q / 10.0 for q in range(1, 10)])
+
+                splits = weighted_quantiles(
+                    self.df[dim + "_encoded"], deciles, self.df["weights"]
+                )
+
+                for split in np.unique(splits):
+                    left = self.df[self.df[dim + "_encoded"] < split]
+                    right = self.df[self.df[dim + "_encoded"] >= split]
+                    if len(left) == 0 or len(right) == 0:
+                        continue
+                    dim_values1 = [k for k, v in enc_map.items() if v < split]
+                    dim_values2 = [k for k, v in enc_map.items() if v >= split]
+                    left_candidate = ModelNode(
+                        df=left,
+                        fitter=self.fitter,
+                        dims=self.dims,
+                        dim_split={**self.dim_split, **{dim: dim_values1}},
+                        depth=self.depth + 1,
+                    )
+                    right_candidate = ModelNode(
+                        df=right,
+                        fitter=self.fitter,
+                        dims=self.dims,
+                        dim_split={**self.dim_split, **{dim: dim_values2}},
+                        depth=self.depth + 1,
+                    )
+
+                    err = left_candidate.error + right_candidate.error
+                    if err < best_error:
+                        best_error = err
+                        self._error_improvement = self.error - best_error
+                        self._best_submodels = (left_candidate, right_candidate)
+
+        return self._error_improvement
+
+
+def mod_improvement(improvement: float, depth: int, max_depth: int) -> float:
+    if depth < max_depth:
+        return improvement
+    else:
+        return float("-inf")
+
+
+def get_best_subtree_result(
+    node: ModelNode, max_depth: Optional[int] = 1000
+) -> ModelNode:
+    if node.children is None or node.depth >= max_depth:
+        return node
+    else:
+        node1 = get_best_subtree_result(node.children[0])
+        node2 = get_best_subtree_result(node.children[1])
+        improvement1 = mod_improvement(node1.error_improvement, node1.depth, max_depth)
+        improvement2 = mod_improvement(node2.error_improvement, node2.depth, max_depth)
+        if improvement1 > improvement2:
+            return node1
+        else:
+            return node2
+
+
+def build_tree(root: ModelNode, num_leaves: int, max_depth: Optional[int] = 1000):
+    # TODO: modify this to also accept max_depth
+    for _ in range(num_leaves - 1):
+        best_node = get_best_subtree_result(root, max_depth)
+        if best_node.error_improvement > 0:
+            best_node.children = best_node._best_submodels
+        else:
+            break
```

Comment:
if we want to end in case of error_improvement <= 0, it may be more stable to do "return" I think

Reply:
Why? In this case I think they're exactly equivalent, no?